woocommerce/woocommerce
# [Experimental] Refactor: replace QueryFilters by ProductQueryFiltersServiceProvider

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:
- Remove `QueryFilters` class from `Blocks`.
- Add a new service provider named `ProductQueryFiltersServiceProvider` which provides two services:
  - `FilterClauses` to generate required posts clauses for filtering.
  - `FilterData` to calculate the data for filter blocks.
- For `FilterData`, I use setter to allow changing the clauses provider at the run time. The reason for that is to refactor StoreAPI later as I try in #43811.
- Adds unit test to cover new classes.




Closes #43493.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Make sure the `blocks.ini` content looks like the following:
```
woocommerce_blocks_phase = 3
woocommerce_blocks_env = development
```

1. Make sure you have some products in your store
2. The expected behaviors for Attribute Filter using the `AND` condition:
    - The count of remaining attributes is updated when selecting one or more terms. Eg, if `Blue (4)` is selected, `Green (3)` will be updated to `Green (2)` because there are two products that have both `Blue` **and** `Green` attributes.
    - The number of filtered products always matches the count of selected terms.
    - The count of selected terms is always the same between each other. Eg, if `Blue (4)` is selected, `Green (3)` will be updated to `Green (2)`. if `Green (2)` is selected then, `Blue (4)` will be `Blue (2)`.
2. The expected behaviors for Attribute Filter using the `OR` condition:
    - The count of remaining attributes doesn't change when selecting one or more terms. Eg, if `Blue (4)` is selected, `Green (3)` will remain as `Green (3)` because we count the single attribute only when `OR` is used.
    - The number of filtered products can be greater than the maximum count of selected attributes, but always smaller than the sum of them. Eg:  we have `Blue (4)` and `Green (3)` but when both are selected, the filtered product may be only 5 because there are two products containing both `Blue` and `Green`
    - The count of selected terms can be different.
3. The condition between attributes is always `AND`. For example, if you have a color and a size filter, with `Blue` and `Large` selected respectively, the filtered products (and count) are products having `Blue` color **and** `Large` size. Selecting a color will update the count for sizes.
 
#### For product archive templates:

1. Go to Site Editor > Templates > Product Catalog.
4. Replace the existing Products (Beta) block with the `Product Collection (Beta)` block. Keep the `Sync with current query` enabled.
5. Add 5 filter blocks to the page: `Product Filter: Active Filter (Beta)`, `Product Filter: Price (Beta)`, `Product Filter: Stock (Beta)`, `Product Filter: Rating (Beta)`, `Product Filter: Attribute (Beta)`.
6. For Attribute, Stock, and Rating blocks, turn on the count from the Inspector setting.
7. Verify new filter blocks work on the shop page. The filter data should match with the current product query.
8. Try different combinations of filters, see blocks always show correct filter data.
9. Try changing the filter condition for the Attribute Filter block, see the counts updated and followed the desired behavior above. 
10. Note that the result count block isn't interactive so it won't update during filtering.

Repeat steps above with other product archive templates:
- Product Category/Tag.
- Product Attribute Archive.
- Product Search Results.

#### For custom Product Collection query. 

1. On a new page, add the `Product Collection` block.
4. Add new beta filter blocks (`Product Filter: Active Filter (Beta)`, `Product Filter: Price (Beta)`, `Product Filter: Stock (Beta)`, `Product Filter: Rating (Beta)`, `Product Filter: Attribute (Beta)`) **inside** the `Product Collection` block (and outside of the `Product Template` block).
5. For Attribute, Stock, and Rating blocks, turn on the count from the Inspector setting.
6. Save and see the new filters work on the front end. The filter data should match with the current product query.
8. Try different combinations of filters, see blocks always show correct filter data.
11. Edit the Product Collection, and add some backend filters like price range, category, tag, or stock status.
12. See the filter counts data updated to match the new product query on the front end.
13. Try changing the filter condition for the Attribute Filter block, see the counts updated and followed the desired behavior above.

#### Test cache invalidating

1. Try visit a test page from steps above.
2. Record the counts.
4. Try updating products to affect the filter results (like change product stock status, add/remove variation with attributes, change the price).
5. See the counts is updated with new results.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
